lib/samvera/derivatives.rb

```diff
@@ -0,0 +1,203 @@
+        end
+      end
+    end
+
+    module FileApplicator
+      ##
+      # @api public
+      #
+      # @param file_set [FileSet]
+      # @param derivative_type [#to_sym]
+      # @param from_location [Object]
+      # @param strategies [Array<#write!>]
+      def self.call(file_set:, derivative_type:, from_location:, strategies: default_strategies_for(file_set))
+        strategies.each do |strategy|
+          strategy.write!(file_set: file_set, derivative_type: derivative_type, from_location: from_location)
+        end
+      end
+
+      # @note By providing a FileSet we are allowing for a strategy to be repurposed for multiple
+      #       derivative types; this would mean that we could wrap and re-use the existing
+      #       Hyrax::FileSetDerivativeService logic as a strategy.
+      def self.default_strategies_for(file_set)
+        return file_set.derivative_applicator_strategies if file_set.respond_to?(:derivative_applicator_strategies)
+        # TODO: What's the default?  A wrapper around Hyrax::FileSetDerivativeService?
+      end
+      private_class_method :default_strategies_for
+
+      ##
+      # @abstract
+      #
+      # The purpose of this abstract class is to provide the public interface for strategies.
+      #
+      # @see {.find}
+      class Strategy
+        # In some cases the FromLocation knows how to write itself; this is the case when we wrap
+        # the Hyrax::FileSetDerivativeService.
+        class_attribute :delegate_write_to_given_from_location, default: false
+
+        ##
+        # @param file_set [FileSet]
+        # @param derivative_type [#to_sym]
+        # @param from_location [Object]
+        def self.write!(file_set:, derivative_type:, from_location:)
+          new(file_set: file_set, derivative_type: derivative_type, from_location: from_location).write!
+        end
+
+        def initialize(file_set:, derivative_type:, from_location:)
+          @file_set = file_set
+          @derivative_type = derivative_type
+          @from_location = from_location
+        end
+        attr_reader :file_set, :derivative_type, :from_location
+
+        # @note What's going on with this logic?  To continue to leverage
+        #       Hyrax::FileSetDerivativeService, we want to let that wrapped service (as a
+        #       FromLocation) to do it's original work.  However, we might have multiple strategies
+        #       in play for application.  That case is when we want to first check for an existing
+        #       thumbnail and failing that generate the thumbnail.  The from_location could either
+        #       be the found thumbnail...or it could be the wrapped Hyrax::FileSetDerivativeService
+        #       that will create the thumbnail and write it to the location.  The two applicator
+        #       strategies in that case would be the wrapper and logic that will write the found
+        #       file to the correct derivative path.
+        def write!
```

Comment:
This may not be the best place for this switching logic; I suspect there's another class that's somewhat itching to emerge.